sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala

```diff
@@ -211,37 +264,69 @@ trait StreamTest extends QueryTest with Timeouts {
+            assert(currentStream != null, "can not stop a stream that is not running")
+            try failAfter(streamingTimout) {
+              currentStream.stop()
+              assert(!currentStream.microBatchThread.isAlive,
+                s"microbatch thread not stopped")
+              assert(!currentStream.isActive,
+                "query.isActive() is false even after stopping")
+              assert(currentStream.exception.isEmpty,
+                "query.exception() is not empty after clean stop")
+            } catch {
+              case _: InterruptedException =>
+              case _: org.scalatest.exceptions.TestFailedDueToTimeoutException =>
+                failTest("Timed out while stopping and waiting for microbatchthread to terminate.")
+              case t: Throwable =>
+                failTest("Error while checking stream failure.", t)
+            } finally {
+              lastStream = currentStream
+              currentStream = null
+            }
 
           case DropBatches(num) =>
-            checkState(currentStream == null, "dropping batches while running leads to corruption")
+            assert(currentStream == null, "dropping batches while running leads to corruption")
             sink.dropBatches(num)
 
-          case ExpectFailure =>
+          case ef: ExpectFailure[_] =>
+            assert(currentStream != null, "dropping batches while running leads to corruption")
             try failAfter(streamingTimout) {
-              while (streamDeathCause == null) {
-                Thread.sleep(100)
+              val thrownException = intercept[QueryException] {
+                currentStream.awaitTermination()
+              }
+              eventually("microbatch thread not stopped after termination with failure") {
+                Assertions.assert(!currentStream.microBatchThread.isAlive)
               }
+              assert(currentStream.exception === Some(thrownException),
+                s"incorrect exception returned by query.exception()")
+
+              val exception = currentStream.exception.get
+              assert(exception.cause.getClass === ef.causeClass,
+                "incorrect cause in exception returned by query.exception()\n" +
+                  s"\tExpected: ${ef.causeClass}\n\tReturned: ${exception.cause.getClass}")
             } catch {
               case _: InterruptedException =>
               case _: org.scalatest.exceptions.TestFailedDueToTimeoutException =>
-                fail(
-                  s"""
-                     |Timed out while waiting for failure.
-                     |$testState
-                   """.stripMargin)
+                failTest("Timed out while waiting for failure.")
+              case t: Throwable =>
+                failTest("Error while checking stream failure", t)
+            } finally {
+              lastStream = currentStream
+              currentStream = null
+              streamDeathCause = null
```

Comment:
I guess I was arguing the contrary.  Since its in a finally I'm trying to reason about why, when in fact its essentially no-op code in the case that an exception is thrown here.